doc/arch/adr-002-form-representation.md

```diff
@@ -0,0 +1,95 @@
+- Users create forms with an arbitrary number of questions.
+- Question forms can have a variety of field types—text, checkbox, radio buttons
+  and dropdown boxes. It should be easy to add more response types later.
+- Users provide an unknown number of responses for each question form. Each
+  response maps to the structure and requirements of its question form.
+- Sufficiently fast reads for retrieving form data and collating all answers in
+  the responses.
+
+The application currently uses PostgreSQL for storing user data.
+
+With that in mind, some options include:
+- Representing form questions and answers in JSON
+  - using PostgreSQL's `json` type for the column having the form data.
+  - using PostgreSQL's `text` type for the column having the form data.
+  - using PostgreSQL's `jsonb` type as the column having the form data.
+  - storing form data in a separate document store
+- Representing and storing form questions and answers in relations
+- Storing the form questions directly as HTML, and storing the answers in a relation
+
+## Decision
+
+We will use PostgreSQL's `jsonb` columns to store form questions and answers.
+The illustration below demonstrates how we will model the form data.
+
+![Data model](https://user-images.githubusercontent.com/24277692/156551390-af11e56a-f8d6-4d7f-a55b-b8d4e511104c.png)
+
+We still use relations for storing form and response IDs, along with their
+metadata. The form structure will be stored in a `jsonb` column.
+
+
+JSON arrays represent a forms and responses. Each item in an array will be an
+object containing a unique (to the form) question ID along with a type field and
+its associated metadata.
+
+Here's an example of a form with two questions:
+```json
+[
+  {
+    "id": "a283fe",
+    "type": "text",
+    "question": "What did you have for oota?"
+  },
+  {
+    "id": "ca92e3",
+    "type": "single-choice",
+    "question": "Pick your favourite language.",
+    "choice": ["Clojure", "C++", "Python", "Gujarati"],
+    "mandatory": true
+  }
+]
+```
+
+The application converts these JSON representation of the questions to HTML
+forms when presented to a responder.
+
+Responses are collected and converted to a JSON representation as well. A sample
+response to this form might look like:
+```json
+[
+  {
+    "id": "a283fe",
+    "type": "text",
+    "answer": "Undhiyu and Rotla"
+  },
+  {
+    "id": "ca92e3",
+    "type": "single-choice",
+    "answer": "Gujarati"
+  }
+]
+```
+
+Note that the ids in responses correspond to the id of the question they answer.
+This simplifies matching a question to an answer.
+
+## Status
+
+Proposed
+
+## Consequences
+
+Using JSON for storing the form questions and responses gives us the flexibility
+that enables us to create forms with arbitrary number of questions and
+responses. It's easy to add more response types without changes to the existing
+data models. Using JSON also decouples the representation from the presentation.
```

Comment:
Are there any pitfalls that we know of using JSON compared to the other approaches? This section should also mention those.

Reply:
Gotcha. Will be part of the tradeoffs section.